lang/src/org/partiql/lang/eval/physical/operators/Accumulator.kt

```diff
@@ -0,0 +1,183 @@
+                "min" -> AccumulatorMin(valueFactory, filter)
+                "max" -> AccumulatorMax(valueFactory, filter)
+                "avg" -> AccumulatorAvg(valueFactory, filter)
+                "count" -> AccumulatorCount(valueFactory, filter)
+                "sum" -> AccumulatorSum(valueFactory, filter)
+                "group_as" -> AccumulatorGroupAs(valueFactory, filter)
+                else -> throw IllegalArgumentException("Unsupported aggregation function: $funcName")
+            }
+        }
+    }
+
+    override fun next(value: ExprValue) {
+        if (value.isUnknown() || filter.invoke(value).not()) return
+        nextValue(value)
+    }
+
+    abstract fun nextValue(value: ExprValue)
+}
+
+internal class AccumulatorSum(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var sum: Number? = null
+
+    override fun nextValue(value: ExprValue) {
+        checkIsNumberType(funcName = "SUM", value = value)
+        if (sum == null) sum = 0L
+        this.sum = value.numberValue() + this.sum!!
+    }
+
+    override fun compute(): ExprValue {
+        return sum?.exprValue(valueFactory) ?: valueFactory.nullValue
+    }
+}
+
+internal class AccumulatorAvg(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var sum: Number = 0.0
+    var count: Long = 0L
+
+    override fun nextValue(value: ExprValue) {
+        checkIsNumberType(funcName = "AVG", value = value)
+        this.sum += value.numberValue()
+        this.count += 1L
+    }
+
+    override fun compute(): ExprValue = when (count) {
+        0L -> valueFactory.nullValue
+        else -> (sum / bigDecimalOf(count)).exprValue(valueFactory)
+    }
+}
+
+internal class AccumulatorMax(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var max: ExprValue = valueFactory.nullValue
+
+    override fun nextValue(value: ExprValue) {
+        max = comparisonAccumulator(NaturalExprValueComparators.NULLS_LAST_DESC)(max, value)
+    }
+
+    override fun compute(): ExprValue = max
+}
+
+internal class AccumulatorMin(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var min: ExprValue = valueFactory.nullValue
+
+    override fun nextValue(value: ExprValue) {
+        min = comparisonAccumulator(NaturalExprValueComparators.NULLS_LAST_ASC)(min, value)
+    }
+
+    override fun compute(): ExprValue = min
+}
+
+internal class AccumulatorCount(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var count: Long = 0L
+
+    override fun nextValue(value: ExprValue) {
+        this.count += 1L
```

Comment:
At least for nulls, Section 11.2 says: "The group-by and aggregation of PartiQL is backwards compatible to SQL" -- where SQL filters nulls with the exception of COUNT(*). We still follow this -- see `Accumulator.next()`

As for MISSING, it doesn't seem the spec elaborates on NULLS and MISSING specifically for aggregate functions. It does talk about it in the context of grouping (related): "Notice that PartiQL will group together the NULL and the MISSING grouping expressions, since any grouping expression resulting to MISSING has been coerced into NULL before eqg does comparisons for grouping."

Beyond that, the Evaluating Compiler filters `NULL` and `MISSING` values (similar to the above mentioned `Accumulator.next()`) while accumulating.

Reply:
Also, to note: for typical SQL, `COUNT(*)` is exempt from NULL values and accumulates for every row encountered. This is still kept with the PR implementation since `COUNT(*)` is parsed into `call('count', lit(1))` -- where the value expression will never evaluate to `NULL`